cli/popper/runner_slurm.py

```diff
@@ -0,0 +1,220 @@
+import os
+import tempfile
+import subprocess
+
+from popper import utils as pu
+from popper import scm
+from popper.cli import log as log
+from popper.runner import StepRunner as StepRunner
+from popper.runner_host import DockerRunner as HostDockerRunner
+from popper.runner_host import HostRunner
+
+
+class SlurmRunner(StepRunner):
+
+    spawned_jobs = set()
+
+    def __init__(self, config):
+        super(SlurmRunner, self).__init__(config)
+
+    def __exit__(self, exc_type, exc, traceback):
+        SlurmRunner.spawned_jobs = set()
+
+    def stream_output(self, job_id):
+        log.debug(f"Attaching to Job {job_id}")
+        while True:
```

Comment:
I cannot quite grasp the need for this loop and the subsequent `sattach` invocation. If this is due to `sattach` failing because the job doesn’t exist yet, then this could potentially loop infinitely when `docker run` returns a non-exit code (when the job was created and immediately exited between loop iterations).

Also, this loop can potentially bombard the slurm server, so we can add use the `SLURM_EXIT_ERROR` environment variable as described [here](https://slurm.schedmd.com/sattach.html) to obtain an arbitrary exit code for when slurm returns a failure (as when the job doesn't exist). In this way, we loop as long as we find this exit code, and we don't need to make a subsequent call to `sattach`

Reply:
@ivotron Yeah, the loop checks whether for the job is whether the job is in pending state or not. As soon as the job changes state from `P` to `R`, `sattach` receives a exit code of 0. This the point we break from the loop. Note, the commands in the script are still not executed.  Now the subsequent exit code tells us whether the commands pass/fail and the action fails accordingly.

Reply:
I see, so then we need to use another mechanism to check the status of the job. The sattach docs imply that the exit code returned is coming from the underlying `srun`, which in this case would correspond to the exit code from `docker run`. Two alternatives:
 1. do what I mentioned above (use `SLURM_EXIT_ERROR` to disambiguate between `docker run` and `sattach` errors; and remove the 2nd invocation to `sattach`).
 2. use `squeue` to check the status of the job, and then run `sattach` once we know is running.

Reply:
Sorry, but what I meant was there's a stage which returns `0` exit code between `sattach failing for the last time with 1 exit code` and `docker run starting may be 1 or 0`. So, theres no chance of overlapping of `sattach` and `docker run` error. That's the reason i am looping and checking for the first `0` and breaking out there. Does that make sense?

Reply:
Let me try it out with `squeue`. looks like it will be easier.

Reply:
@ivotron looks like the complexity is the same or increases to some extent on using `squeue` to check job status. Even if we use `SLURM_EXIT_ERROR` environment variable, we still need to distinguish between the `0` exit code from `sattach` and the `0` exit code from `docker run`, thus we break and call `sattach` again. The `sattach` command isn't much user friendly, part of which is leading to this increased complexity. I really can't think of any other way to optimize this.